examples/solidity/GradualPonzi/gradualPonzi.qnt

```diff
@@ -0,0 +1,225 @@
+// -*- mode: Bluespec; -*-
+/**
+ * A more elaborate example of a Ponzi scheme, inspired by the Solidity contract [2].
+ * Do not use it in production! The purpose of this example is to show how to
+ * specify and check interesting safety and liveness properties with Quint.
+ *
+ * [1]: https://github.com/djokicx/sol-simple-ponzi/blob/12474f468f4823a39276cac5ef7185b349f5c907/contracts/GradualPonzi.sol
+ *
+ * Igor Konnov, Informal Systems, 2023
+ */
+module gradualPonzi {
+  pure val MINIMUM_INVESTMENT = 10^15
+
+  // addresses are string literals
+  type Addr = str
+
+  // the set of addresses is not fixed, see simplePonziTest below
+  const ADDR: Set[Addr]
+
+  // a state of the EVM that is observed/modified by the contract
+  type EvmState = {
+    // the account balances for every address
+    balances: Addr -> int
+  }
+
+  // a state of the contract
+  type PonziState = {
+    // the address of all previous investors, potentially, with duplicates
+    investors: List[Addr],
+    // rewards distributed so far
+    rewards: Addr -> int,
+  }
+
+  // create a new instance of the contract
+  pure def newPonzi(sender: Addr): PonziState = {
+    investors: [ sender ],
+    rewards: ADDR.mapBy(a => 0),
+  }
+
+  // The result of applying a method.
+  // If error != "", then the states are not modified,
+  // and error contains the error message.
+  // Otherwise, evm and ponzi contain the new states.
+  type Result = {
+    error: str,
+    evm: EvmState,
+    ponzi: PonziState,
+  }
+
+  // Receive an investment and save the rewards (to the previous investors).
+  pure def receive(evm: EvmState,
+      ponzi: PonziState, investor: Addr, amount: int): Result = {
+    if (amount > evm.balances.get(investor)) {
+      error: "Insufficient funds",
+      evm: evm,
+      ponzi: ponzi,
+     } else if (amount < MINIMUM_INVESTMENT) {
+      error: "Investment below minimum",
+      evm: evm,
+      ponzi: ponzi,
+    } else
+      pure val investorCut = amount / length(ponzi.investors)
```

Comment:
Can this integer division be changing the total supply?

Reply:
I have added `noLeftoversInv`, which shows that the contract may accumulate leftover tokens due to rounding. It's actually quite nice.

Reply:
Awesome, makes sense to me!